service/controller/v5/resource/instance/deployment.go

```diff
@@ -35,34 +61,64 @@ func (r Resource) newDeployment(ctx context.Context, obj providerv1alpha1.AzureC
 		workerNodes = append(workerNodes, n)
 	}
 
-	cc, err := controllercontext.FromContext(ctx)
+	containerName := key.BlobContainerName()
+	groupName := key.ResourceGroupName(obj)
+	storageAccountName := key.StorageAccountName(obj)
+	masterBlobName := key.BlobName(obj, key.PrefixMaster())
+	workerBlobName := key.BlobName(obj, key.PrefixWorker())
+
+	storageAccountsClient, err := r.getStorageAccountsClient(ctx)
 	if err != nil {
 		return azureresource.Deployment{}, microerror.Mask(err)
 	}
-	err = cc.Validate()
+
+	keys, err := storageAccountsClient.ListKeys(ctx, groupName, storageAccountName)
 	if err != nil {
 		return azureresource.Deployment{}, microerror.Mask(err)
 	}
 
-	masterCloudConfig, err := cc.CloudConfig.NewMasterCloudConfig(obj)
+	if len(*(keys.Keys)) == 0 {
+		return azureresource.Deployment{}, microerror.Maskf(executionFailedError, "storage account key's list is empty")
+	}
+	primaryKey := *(((*keys.Keys)[0]).Value)
```

Comment:
This is to fill my knowledge gaps. What are those keys? Why the first one is primary? What are other keys? Does it make sense to create `storageAccountsClient.GetPrimaryKey` method?

Reply:
Keys are api keys for accessing storage account.
I'd love to skip extending basic azure client. I don't see we're doing that anywhere. I can move it to separate function, but that's not used that often

Reply:
How about replacing `ListKeys` with `GetPrimaryKey` instead of extending the object?
Will we use `ListKeys` anywhere else?

Reply:
Scratch that. I just realized this is from the azure library.

Reply:
Can we encapsulate getting the primary key in `GetBlobURL` functionality? I think we could just replace `groupName` with `groupName` and do everything there.

Reply:
We can, but then we will call keys listing two times

Reply:
Also I don't passing storage account client into blob_client or retrieving storage account again in blob_client